# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [12]:
import pandas as pd

df = pd.read_csv(
    "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")
df

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

In [ ]:
# It contains 4008 rows over 11 columns
df.shape

In [6]:
"""
8/11 columns are of datatype object; 3/11 are float.
Suggesting to change the dtype:
- Customer Lifetime Value -> float,
- Income -> int,
- Monthly Premium Auto -> int,
- Number of Open Complaints -> int
"""
df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [14]:
"""
Categorical columns:
- Customer,
- ST,
- Gender,
- Education,
- Policy Type,
- Vehicle Class.
"""
df.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [25]:
df.describe(include="all")

Customer      ST GENDER Education Customer Lifetime Value  \
count      1071    1071    954      1071                    1068   
unique     1071       8      5         6                    1027   
top     RB50392  Oregon      F  Bachelor              445811.34%   
freq          1     320    457       324                       4   
mean        NaN     NaN    NaN       NaN                     NaN   
std         NaN     NaN    NaN       NaN                     NaN   
min         NaN     NaN    NaN       NaN                     NaN   
25%         NaN     NaN    NaN       NaN                     NaN   
50%         NaN     NaN    NaN       NaN                     NaN   
75%         NaN     NaN    NaN       NaN                     NaN   
max         NaN     NaN    NaN       NaN                     NaN   

              Income  Monthly Premium Auto Number of Open Complaints  \
count    1071.000000           1071.000000                      1071   
unique           NaN                   NaN                         6   
top              NaN                   NaN                    1/0/00   
freq             NaN                   NaN                       830   
mean    39295.701214            193.234360                       NaN   
std     30469.427060           1601.190369                       NaN   
min         0.000000             61.000000                       NaN   
25%     14072.000000             68.000000                       NaN   
50%     36234.000000             83.000000                       NaN   
75%     64631.000000            109.500000                       NaN   
max     99960.000000          35354.000000                       NaN   

          Policy Type  Vehicle Class  Total Claim Amount  
count            1071           1071         1071.000000  
unique              3              6                 NaN  
top     Personal Auto  Four-Door Car                 NaN  
freq              780            576                 NaN  
mean              NaN            NaN          404.986909  
std               NaN            NaN          293.027260  
min               NaN            NaN            0.382107  
25%               NaN            NaN          202.157702  
50%               NaN            NaN          354.729129  
75%               NaN            NaN          532.800000  
max               NaN            NaN         2893.239678

**Description of the unique values of each categorical column**, which one repeats the most, and how many times:
- Customer: each of 1071 entries/customers has its unique Customer ID.
- ST: has 8 unique values; Oregon repeats the most, 320 times.
- GENDER: 5 unique values, which is unexpected and probably due to non-standardized data entry; "F" has the highest frequency, 457.
- Education: 6 unique values; Bachelor, 324 times.
- Policy Type: 3 unique policies; Personal Auto, 780 times.
- Vehicle Class: 6 unique classes; has Four-Door Car, 576 times

**Range of values for numerical columns**:
- Income: ranges from 0 to 99960
- Monthly Premium Auto: from 61 to 35354
- Total Claim Amount: 0.382107 - 2893.239678

**Summary statistics**:
- Income: it should be clarified in the column name whether it's monthly or annual income.
Median salary is 36, lower than the mean (39), and far from the max (99), suggesting that the income is on the lower end.

- Monthly Premium Auto: it needs to be clarified whether the values in this column are per single vehicle, or they could also be for multiple vehicles belonging to a single customer.
35354 seems extremely high considering that 3/4 of the Premiums are below 109.

- Total Claim Amount: median is lower than the mean claim suggesting that the claims are on the lower end, with some very high claims skewing the distribution to the right.

**Remarks**:
- The following columns are in an inappropriate format (object), so I can't retrieve the summary statistics or range of values:
  - Customer Lifetime Value,
  - Number of Open Complaints.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [43]:
df.ST.value_counts().sort_values().head()

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [60]:
df["Policy Type"].value_counts().idxmax()

'Personal Auto'

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [86]:
# Filter dataframes using loc
df_personal_auto = df.loc[df["Policy Type"] == "Personal Auto"]
df_corporate_auto = df.loc[df["Policy Type"] == "Corporate Auto"]

# Calculate the average income for each policy type
personal_auto_mean_income = df_personal_auto["Income"].mean()
corporate_auto_mean_income = df_corporate_auto["Income"].mean()

# Print the results
print("Personal auto mean income:", personal_auto_mean_income)
print("Corporate auto mean income:", corporate_auto_mean_income)

Personal auto mean income: 38180.69871794872
Corporate auto mean income: 41390.31196581197


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [101]:
df_filtered = df[df["Total Claim Amount"] >
                 df["Total Claim Amount"].quantile(.75)]
df_filtered

Customer          ST GENDER Education Customer Lifetime Value   Income  \
1     QZ44356     Arizona      F  Bachelor              697953.59%      0.0   
2     AI49188      Nevada      F  Bachelor             1288743.17%  48767.0   
17    OE15005        Cali    NaN   College              394524.16%  28855.0   
23    TZ98966      Nevada    NaN  Bachelor              245019.10%      0.0   
26    US89481  California    NaN  Bachelor              394637.21%      0.0   
...       ...         ...    ...       ...                     ...      ...   
1059  YG44474      Oregon      M   College             1401472.13%  54193.0   
1061  RY92647        Cali      F  Bachelor             1050677.17%      0.0   
1068  GS98873     Arizona      F  Bachelor              323912.47%  16061.0   
1069  CW49887  California      F    Master              462680.11%  79487.0   
1070  MY31220  California      F   College              899704.02%  54230.0   

      Monthly Premium Auto Number of Open Complaints     Policy Type  \
1                     94.0                    1/0/00   Personal Auto   
2                    108.0                    1/0/00   Personal Auto   
17                   101.0                    1/0/00   Personal Auto   
23                    73.0                    1/3/00  Corporate Auto   
26                   111.0                    1/0/00   Personal Auto   
...                    ...                       ...             ...   
1059                 117.0                    1/0/00  Corporate Auto   
1061                  92.0                    1/0/00   Personal Auto   
1068                  88.0                    1/0/00   Personal Auto   
1069                 114.0                    1/0/00    Special Auto   
1070                 112.0                    1/0/00   Personal Auto   

      Vehicle Class  Total Claim Amount  
1     Four-Door Car         1131.464935  
2      Two-Door Car          566.472247  
17              SUV          647.442031  
23    Four-Door Car          554.376763  
26    Four-Door Car          799.200000  
...             ...                 ...  
1059            SUV          720.752945  
1061  Four-Door Car          546.524896  
1068  Four-Door Car          633.600000  
1069            SUV          547.200000  
1070   Two-Door Car          537.600000  

[264 rows x 11 columns]

In [103]:
df_filtered.describe()

Income  Monthly Premium Auto  Total Claim Amount
count    264.000000            264.000000          264.000000
mean   23677.344697            165.193182          782.228263
std    27013.483721            623.930992          292.751640
min        0.000000             63.000000          537.600000
25%        0.000000             99.000000          606.521741
50%    18807.000000            114.000000          679.597985
75%    42423.750000            133.250000          851.400000
max    99316.000000          10202.000000         2893.239678